HW2/P2/parallel_vector.pyx

```diff
@@ -18,98 +19,115 @@ cimport numpy as np
-
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for each element of counts to parallelize
-   # data movement.  Be sure to avoid deadlock, and double-locking.
-   ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
-
-   free_N_locks(counts.shape[0], locks)
-
+							 np.int32_t[:] src,
+							 np.int32_t[:] dest,
+							 int repeat):
+	cdef:
+		int idx, r
+		omp_lock_t *locks = get_N_locks(counts.shape[0])
+	
+	##########
+	# Your code here
+	# Use parallel.prange() and a lock for each element of counts to parallelize
+	# data movement.  Be sure to avoid deadlock, and double-locking.
+	##########
+	with nogil:
+		for r in range(repeat):
+			for idx in prange(src.shape[0],num_threads=4):
+				acquire(&locks[src[idx]])
+				if counts[src[idx]] > 0:	
+					counts[src[idx]] -= 1
+					release(&locks[src[idx]])
+					acquire(&locks[dest[idx]])
+					counts[dest[idx]] += 1
+					release(&locks[dest[idx]])
+				else:
+					#Release so there is no deadlock
+					release(&locks[src[idx]])
+		
+
+	free_N_locks(counts.shape[0], locks)
 
 cpdef move_data_medium_grained(np.int32_t[:] counts,
-                               np.int32_t[:] src,
-                               np.int32_t[:] dest,
-                               int repeat,
-                               int N):
-   cdef:
-       int idx, r
-       int num_locks = (counts.shape[0] + N - 1) / N  # ensure enough locks
-       omp_lock_t *locks = get_N_locks(num_locks)
-
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for every N adjacent elements of counts
-   # to parallelize data movement.  Be sure to avoid deadlock, as well as
-   # double-locking.
-   ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
-
-   free_N_locks(num_locks, locks)
+							np.int32_t[:] src,
+							np.int32_t[:] dest,
+							int repeat,
+							int N):
+	cdef:
+		int idx, r
+		int num_locks = (counts.shape[0] + N - 1) / N  # ensure enough locks
+		#int num_locks = N  # ensure enough locks
+		omp_lock_t *locks = get_N_locks(num_locks)
+		int divider = counts.shape[0]/N
+	##########
+	# Your code here
+	# Use parallel.prange() and a lock for every N adjacent elements of counts
+	# to parallelize data movement.  Be sure to avoid deadlock, as well as
+	# double-locking.
+	##########
+	with nogil:
+		for r in range(repeat):
+			for idx in prange(src.shape[0],num_threads=4):
+				acquire(&locks[src[idx]/N])
+				if counts[src[idx]] > 0:
+					counts[src[idx]] -= 1
+					release(&locks[src[idx]/N])
+					acquire(&locks[dest[idx]/N])
+					counts[dest[idx]] += 1
+					release(&locks[dest[idx]/N])
+				else:
```

Comment:
im curious as well. And would you need to check which one to acquire on line 121 before you choose the src ?